opacus/grad_sample/linear.py

```diff
@@ -21,7 +21,11 @@ def compute_linear_grad_sample(
         activations: Activations
         backprops: Backpropagations
     """
-    gs = torch.einsum("n...i,n...j->nij", backprops, activations)
+    # With activations with 3 dimensions (transformers) the original formula does not aggregate correctly
+    if activations.ndim == 2:
+        gs = torch.einsum("n...i,n...j->nij", backprops, activations)
+    else:
```

Comment:
Noob question: I thought ellipses in einsum handles all dimensions, and by extension even 3D. Do you know why it didn't work?

Reply:
After careful investigation today, it appears that its linked to a previous version fo PyTorch that we used in our environment (specifically `1.7.1`). The bug appears to be solved for `1.9` since we could not reproduce for this version.

My suggestion would then be either (1) not to merge the PR but update the requirements for PyTorch or (2) merge the PR so that we have backward compatibility with older PyTorch versions.

![image](https://user-images.githubusercontent.com/9300642/140325983-5a9cd207-8cbd-4a03-a13a-3f068edb593d.png)

Reply:
Updating PyTorch requirements past 1.8 is problematic, because we depend on csprng, which doesn't support 1.9 and beyond.
Is the bug still reproducable in  1.8?

Reply:
Good news, bug disappears for 1.8.0. Let's update the requirements then?

![image](https://user-images.githubusercontent.com/9300642/140554038-bd28b0d0-bee6-4276-a878-055e5949148e.png)

Reply:
> Let's update the requirements then?

+1

Reply:
@ffuuugor Are you sure csprng doesn't support torch-1.9 (in it's current state). Did you see [my comment](https://github.com/pytorch/csprng/issues/130)? Just checking. I didn't test it deeply.

Reply:
You're probably right - but I haven't looked too deeply either.
I was only relying on the fact that PyPI release depend on 1.8 and their documentation explicitly lists all pyTorch versions they support, notably lacking 1.9 and beyond.
Is the fix as simple as installing from main branch instead of the latest release?